vpr/src/analytical_place/read_atom_netlist.cpp

```diff
@@ -0,0 +1,175 @@
+            int blk_x_loc = region_rect.xmin();
+            int blk_y_loc = region_rect.ymin();
+            // Get the layer.
+            VTR_ASSERT(region.get_layer_range().first == region.get_layer_range().second && "Expect each region to be a single point in layer!");
+            int blk_layer_num = region.get_layer_range().first;
+            // Get the sub_tile (if fixed).
+            int blk_sub_tile = APFixedBlockLoc::UNFIXED_DIM;
+            if (region.get_sub_tile() != NO_SUBTILE)
+                blk_sub_tile = region.get_sub_tile();
+            // Set the fixed block location.
+            APFixedBlockLoc loc = {blk_x_loc, blk_y_loc, blk_layer_num, blk_sub_tile};
+            ap_netlist.set_block_loc(ap_blk_id, loc);
+        }
+    }
+
+    // Cleanup the netlist by removing undesirable nets.
+    // Currently undesirable nets are nets that are:
+    //  - ignored for placement
+    //  - a global net
+    //  - connected to 1 or fewer unique blocks
+    //  - connected to only fixed blocks
+    // TODO: Allow the user to pass a threshold so we can remove high-fanout nets.
+    auto remove_net = [&](APNetId net_id) {
+        // Remove all pins associated with this net
+        for (APPinId pin_id : ap_netlist.net_pins(net_id))
+            ap_netlist.remove_pin(pin_id);
+        // Remove the net
+        ap_netlist.remove_net(net_id);
+    };
+    for (APNetId ap_net_id : ap_netlist.nets()) {
+        // Is the net ignored for placement, if so remove
+        const std::string& net_name = ap_netlist.net_name(ap_net_id);
+        AtomNetId atom_net_id = atom_netlist.find_net(net_name);
+        VTR_ASSERT(atom_net_id.is_valid());
+        if (atom_netlist.net_is_ignored(atom_net_id)) {
+            remove_net(ap_net_id);
+            continue;
+        }
+        // Is the net global, if so remove
+        if (atom_netlist.net_is_global(atom_net_id)) {
+            remove_net(ap_net_id);
+            continue;
+        }
+        // Get the unique blocks connectioned to this net
+        std::unordered_set<APBlockId> net_blocks;
+        for (APPinId ap_pin_id : ap_netlist.net_pins(ap_net_id)) {
+            net_blocks.insert(ap_netlist.pin_block(ap_pin_id));
+        }
+        // If connected to 1 or fewer unique blocks, remove
+        if (net_blocks.size() <= 1) {
+            remove_net(ap_net_id);
+            continue;
+        }
+        // If all the connected blocks are fixed, remove
+        bool is_all_fixed = true;
+        for (APBlockId ap_blk_id : net_blocks) {
+            if (ap_netlist.block_mobility(ap_blk_id) == APBlockMobility::MOVEABLE) {
+                is_all_fixed = false;
+                break;
+            }
+        }
+        if (is_all_fixed) {
+            remove_net(ap_net_id);
+            continue;
+        }
+    }
+    ap_netlist.remove_and_compress();
+
+    // TODO: Some nets may go to the same block mutliple times. Should that be
```

Comment:
I'd allow nets to go to the same block multiple times. It can happen (e.g. data and clear of a FF, sign extension, etc.) and the timing won't necessarily be the same (some could go to registers and some to internal combinational paths). So when we optimize timing having both connections is probably cleaner.